scipy/spatial/src/function_ref.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include <type_traits>
+
+// Type-erased reference to a function object.
```

Comment:
I feel a bit worried including such low-level infrastructure code in a subproject. The usual maintenance concerns may apply.

I yet have to digest the full context of the PR. Speaking fast, would templates achieve the same goal as a type erased reference to a function object?

Reply:
> would templates achieve the same goal as a type erased reference to a function object?

Yes, I could template `pdist`/`cdist` on the metric type. That's okay, but leads to greater binary size and compile times.

Reply:
And performance is similar because the jump is before the loops?
In cKDTree we have similar metric functions but they are called per data point; so the calculation is a bit different. Have you put any thoughts into consolidating the metric implemenation in both places?

Reply:
>And performance is similar because the jump is before the loops?

Exactly. Function pointers are slow because they prevent inlining and all the optimizations that come from having the code visible to the compiler. Since the tight loops are all in the function, the compiler can optimize them fully.

> In cKDTree we have similar metric functions but they are called per data point; so the cost-gain calculation is a bit different. Have you put any thoughts into consolidating the metric implemenation in both places?

We could include `distance_metrics.h` in ckdtree and call the same metrics. No function pointer indirection is required, that's entirely seperate.

Although as you say, we may want them to be optimized differently.

Reply:
I am not sure this is the place to discuss ckdtree, but there is a design choice that came from @aarchiba’s pyrex code. The build function does not shuffle the rows in the input array. When the query functions hit a leaf node, they cannot run a tight loop over a contiguous block of data. The data points are where they used to be. So they have to use the index array to jump around and fetch data from RAM to collect the points in the leaf. Since a fetch from RAM is about 100 ns and a fetch from L1 is about 0.5 ns, it is not really efficient. I put in all those prefetch statements in the query code to partially remedy this. However, it is still a lot slower than doing a query on a shuffled data array. If we did shuffle the data, however, then maximal optimisation would require code similar to what you have in this PR. The advantage of @aarchiba’s design is that it saves memory. Instead of having a local copy of the data array we get away with storing an array of indices. But is does have performance consequences and they relate to what you are doing in this PR.

Reply:
The good news is that ckdtree hides this complexity from the Python interface, so we could actually change it, but it would require a major overhaul.